test/e2e/scheduling/taint_based_evictions.go

```diff
@@ -0,0 +1,257 @@
+				NodeName: nodeName,
+				Tolerations: []v1.Toleration{
+					{
+						Key:               schedulerapi.TaintNodeUnreachable,
+						Operator:          v1.TolerationOpExists,
+						Effect:            v1.TaintEffectNoExecute,
+						TolerationSeconds: &tolerationSeconds,
+					},
+				},
+				DeletionGracePeriodSeconds: &zero,
+			})
+		}
+		// pod2 is a normal pod without any toleration
+		pods[2] = createPausePod(f, pausePodConfig{
+			Name:     fmt.Sprintf("%v-%v", base, 2),
+			NodeName: nodeName,
+		})
+
+		By("verify all pods are running properly")
+		for _, pod := range pods {
+			framework.ExpectNoError(framework.WaitForPodRunningInNamespace(cs, pod))
+		}
+
+		stopCh := make(chan struct{})
+		podUpdated, podDeleted := make(chan *v1.Pod), make(chan *v1.Pod)
+		By("watching all pods")
+		runTBEController(cs, podUpdated, podDeleted, stopCh, ns)
+
+		// get the node API object
+		nodeSelector := fields.OneTermEqualSelector("metadata.name", nodeName)
+		nodeList, err := cs.CoreV1().Nodes().List(metav1.ListOptions{FieldSelector: nodeSelector.String()})
+		if err != nil || len(nodeList.Items) != 1 {
+			framework.Failf("expected no err, got %v; expected len(nodes) = 1, got %v", err, len(nodeList.Items))
+		}
+		node := nodeList.Items[0]
+
+		By(fmt.Sprintf("Block traffic from node %s to the master", nodeName))
+		host, err := framework.GetNodeExternalIP(&node)
+		// TODO(Huang-Wei): make this case work for local provider
+		// if err != nil {
+		// 	host, err = framework.GetNodeInternalIP(&node)
+		// }
+		framework.ExpectNoError(err)
+		master := framework.GetMasterAddress(cs)
+		taint := newUnreachableNoExecuteTaint()
+
+		defer close(stopCh)
+		defer func() {
+			By(fmt.Sprintf("Unblock traffic from node %s to the master", node.Name))
+			framework.UnblockNetwork(host, master)
+
+			if CurrentGinkgoTestDescription().Failed {
+				return
+			}
+
+			By(fmt.Sprintf("Expect to see node %q becomes Ready", nodeName))
+			framework.WaitForNodeToBeReady(cs, nodeName, time.Minute*1)
+			By("Expect to see unreachable=:NoExecute taint is taken off")
+			if err := framework.WaitForNodeHasTaintOrNot(cs, nodeName, taint, false, time.Second*30); err != nil {
+				framework.Failf(err.Error())
+			}
+		}()
+
+		framework.BlockNetwork(host, master)
+
+		By(fmt.Sprintf("Expect to see node %q becomes NotReady", nodeName))
+		if !framework.WaitForNodeToBeNotReady(cs, nodeName, time.Minute*3) {
+			framework.Failf("node %q doesn't turn to NotReady after 3 minutes", nodeName)
+		}
+		By("Expect to see unreachable=:NoExecute taint is applied")
+		if err := framework.WaitForNodeHasTaintOrNot(cs, nodeName, taint, true, time.Second*30); err != nil {
+			framework.Failf(err.Error())
+		}
+
+		By("Wait 30 seconds so that taint based evictions related behavior can happen")
+		// a) expect pod0 to be evicted immediately
+		// b) expect pod1 to remain the same
+		// c) expect pod2 to be updated with a toleration with tolerationSeconds=300
+		for {
```

Comment:
```suggestion
		for numSeenEvents := 0; numSeenEvent < 3; {
```

Reply:
As I said above, I would much prefer a more deterministic approach where we wait for specific conditions to change.